### PR TITLE
Add Examples page to the Welcome Screen

### LINKED ISSUE #3096: Add "browsing examples" section to the Welcome Screen
The "browsing examples" section is currently missing from the Welcome Screen:

image

image

### Goals for the first iteration

- The example list "belongs" to the Welcome Screen (so it's available when the Welcome Screen is, which currently means "as background", when all recordings are closed).
- Example list is built based on a hosted json file containing example metadata.
- Only list examples for which we have a built RRD file (same as with the current [demo.rerun.io](https://demo.rerun.io)).
- Display: simple list of example summaries including title, description, tags, and thumbnail.
- Launching an example triggers downloading the RRD from web.
- While downloading the RRD, the existing "loading screen" shows up.

### Skip for later

- search bar UI
- filter by tag UI
- download progress bar

### TODO

- [x] https://github.com/rerun-io/rerun/issues/3105
- [x] support for launching streams: https://github.com/rerun-io/rerun/issues/2229
- [ ] nice "example card UI"
- [ ] lazy-loaded thumbnails using https://github.com/emilk/egui/issues/3291

### PR DESCRIPTION
**Pull request by abey79**
### What

Add Example page to the Welcome Screen.

>[!NOTE]
>~~This PR bakes in the example thumbnails, which will incur a significant bump in the wasm build. We will remove this once we pull thumbnails from the web using emilk/egui#3291.~~

Fixes #3096 

image

### TODO

- [x] fix layout issues
- [x] display tags
- [x] have dedicated, short copy for the description: #3201

### Not included in this PR

- **WARNING**: here, we bake in a manifest with hard-coded links to RRDs that were generated within this PR. This will lead to issue down the line, when the RRD format changes.
  - https://github.com/rerun-io/rerun/issues/3212
  - #3213
- download updated manifest
  -  #3190 
- load thumbnail from the web
  - emilk/egui#3291
- provide feedback while downloading a RRD
  - https://github.com/rerun-io/rerun/issues/3192

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/3191) (if applicable)

- [PR Build Summary](https://build.rerun.io/pr/3191)
- [Docs preview](https://rerun.io/preview/3be107e4cc6aa6758a3f22c27a79233b33f2ea6b/docs) 
- [Examples preview](https://rerun.io/preview/3be107e4cc6aa6758a3f22c27a79233b33f2ea6b/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

**Comment by emilk** [comment]
I looks amazing 😍 

But clicking one feels extremely unresponsive, as there is no indication that your click came through. As you noted before, we need to add a load-screen asap. It should be as easy as tweaking the `should_show_welcome_screen` heuristic…

**Comment by abey79** [comment]
@emilk I gave it a quick try but.... it's not that simple (I got all sorts of weird behaviour whose fixing feels beyond the scope of this PR). For the purpose of this PR, I'd like to give some kind of visual feedback (some flashing?) to indicate that the click has been registered, and address the progress feedback in a follow-up PR.

**Comment by martenbjork** [comment]
Almost ready to go! 🚀

Here are some tiny tweaks:

Frame 33

## Typography notes

- It's important to add line height to all text elements, even when they are single line. Without that, the margin calculations become incorrect. An element with a tiny line height looks closer to the element below than an element with a huge line height).
- The heading sizes and font sizes used in this view should be the same as on the welcome screen. That is:

#### Headlines (Connect to live data, Rust, Python, Configure your views, [example name])
```
font-size: 15px
line-height: 22px
font-weight: 550 
```

#### Body text
```
font-size: 13px
line-height: 19px
font-weight: 490 (this will be the new base font weight for the entire app)
```

**Comment by abey79** [comment]
@martenbjork thanks for the detailed feedback—will try to implement as much as possible.

On font weight: I don't believe egui currently has support for font weight (https://github.com/emilk/egui/issues/3218), so we're stuck with whatever is the default.

**Comment by abey79** [comment]
Also, I understand the "copy to tags" vertical space and the "32px vertical gap" to be _minimal values_, as the effective gap depends on the corresponding thumbnail aspect ratio (tags are grid aligned too).

**Comment by martenbjork** [comment]
> Also, I understand the "copy to tags" vertical space and the "32px vertical gap" to be minimal values, as the effective gap depends on the corresponding thumbnail aspect ratio (tags are grid aligned too).

Yes! 💯

File: crates/re_viewer/src/ui/welcome_screen/example_page.rs [hunk 1]
```diff
@@ -0,0 +1,391 @@
+use egui::{NumExt, Ui};
+use re_log_types::LogMsg;
+use re_smart_channel::ReceiveSet;
+use re_ui::ReUi;
+use re_viewer_context::SystemCommandSender;
+use std::collections::HashMap;
+
+// TODO(emilk/egui#3291): replace this by loading image from the web
```
Comment (emilk) [inline crates/re_viewer/src/ui/welcome_screen/example_page.rs:8]: We're very close to merging https://github.com/emilk/egui/pull/3297

File: crates/re_viewer/src/ui/welcome_screen/example_page.rs [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+    url: String,
+    width: u32,
+    height: u32,
+}
+
+#[derive(Debug, serde::Deserialize)]
+#[allow(unused)]
+struct ExampleDesc {
+    name: String,
+    title: String,
```
Comment (emilk) [inline crates/re_viewer/src/ui/welcome_screen/example_page.rs:78]: what's the difference?

File: crates/re_viewer/src/ui/welcome_screen/example_page.rs [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+    rx: &ReceiveSet<LogMsg>,
+    example: &ExampleDesc,
+    size: egui::Vec2,
+    hovered: bool,
+) {
+    // TODO(emilk/egui#3291): pull from web rather than cache
+    let file_name = example.thumbnail.url.split('/').last();
+
+    if let Some(file_name) = file_name {
+        if let Some(icon) = THUMBNAIL_CACHE.get(file_name) {
```
Comment (emilk) [inline crates/re_viewer/src/ui/welcome_screen/example_page.rs:316]: style: can use let-else for less right-ward drift 🤷

File: crates/re_viewer/src/ui/welcome_screen/mod.rs [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    Examples,
+}
+
+// used for the top-level scroll area ID
+impl Hash for WelcomeScreenPage {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        "welcome_screen_page".hash(state);
+        std::mem::discriminant(self).hash(state);
+    }
+}
```
Comment (emilk) [inline crates/re_viewer/src/ui/welcome_screen/mod.rs:24]: Why do you implement this manually?

**Review by emilk — APPROVED** [review]
I love it, but I suggest we wait for https://github.com/emilk/egui/pull/3297 to be merged (any minute now) and the update the PR to use it.

File: crates/re_viewer/src/ui/welcome_screen/example_page.rs (hunk 2 above)
Comment (abey79) [reply crates/re_viewer/src/ui/welcome_screen/example_page.rs:78]: Name is a snake-case identifier of sort, and title is the actual user-facing string. I need to stick to the JSON conventions here (can updated that in another PR), but I'll add some clarifying comments.

File: crates/re_viewer/src/ui/welcome_screen/mod.rs (hunk 4 above)
Comment (abey79) [reply crates/re_viewer/src/ui/welcome_screen/mod.rs:24]: cleaned that up. who knows what the heck went through my mind...

**Comment by jprochazk** [comment]
https://github.com/emilk/egui/pull/3297 has been merged, and this would be the perfect PR to test it on!

The new API is:

```rust

// during setup
egui_extras::loaders::install(&egui_ctx);

// when you want to load an image
ui.image2("http://link.to.image.com/image.png");

```

**Comment by abey79** [comment]
@jprochazk I'll get on it this afternoon.

File: crates/re_viewer/src/ui/welcome_screen/example_page.rs (hunk 1 above)
Comment (abey79) [reply crates/re_viewer/src/ui/welcome_screen/example_page.rs]: done

File: crates/re_viewer/src/ui/welcome_screen/example_page.rs [hunk 5]
```diff
@@ -72,13 +12,19 @@ struct ExampleThumbnail {
 }
 
 #[derive(Debug, serde::Deserialize)]
-#[allow(unused)]
 struct ExampleDesc {
+    // snake_case version of the example name
     name: String,
+
+    // human readable version of the example name
```
Comment (emilk) [inline crates/re_viewer/src/ui/welcome_screen/example_page.rs]:
```suggestion
    /// snake_case version of the example name
    name: String,

    /// human readable version of the example name
```

File: crates/re_viewer/src/ui/welcome_screen/mod.rs [hunk 6]
```diff
@@ -78,11 +70,11 @@ impl WelcomeScreen {
             matches!(self.current_page, WelcomeScreenPage::Welcome),
             true,
         ])
-        .id_source(&self.current_page)
+        .id_source(egui::Id::new("welcome_screen_page").with(&self.current_page))
```
Comment (emilk) [inline crates/re_viewer/src/ui/welcome_screen/mod.rs:73]: Anything that implements `Hash` will work here, including:

```suggestion
        .id_source(("welcome_screen_page", &self.current_page))
```

**Review by emilk — APPROVED** [review]
Awesome!

**Pull request merged**